File: backend/app/schemas/monitor.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.dataset_query import AggregationOp

BaselineStrategy = Literal["previous_period", "moving_average", "zscore"]
CheckFrequency = Literal["hourly", "daily", "manual"]
Direction = Literal["increase", "decrease"]
DetectionMethod = Literal["percentage_change", "moving_average", "zscore"]
Severity = Literal["low", "medium", "high"]
# Mirrors app.monitoring.detection.DetectionStatus — the deterministic
# engine's outcome vocabulary, never decided by the AI (see
# docs/V2_ROADMAP.md's "Natural Language to SQL" architectural principle,
# applied here too: the LLM explains, it never decides).
MonitorStatus = Literal["NO_ANOMALY", "ANOMALY_DETECTED", "INSUFFICIENT_DATA", "ERROR"]


class MonitorCreate(BaseModel):
    dataset_id: UUID
    name: str = Field(min_length=1, max_length=255)
    metric_column: str = Field(min_length=1)
    aggregation: AggregationOp = "sum"
    time_column: str = Field(min_length=1)
    baseline_strategy: BaselineStrategy = "moving_average"
    baseline_window: int = Field(default=7, ge=3, le=90)
    threshold_percent: float | None = Field(default=20.0, gt=0)
    zscore_threshold: float | None = Field(default=3.0, gt=0)
    check_frequency: CheckFrequency = "daily"
    notify_email: bool = False


class MonitorUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    baseline_strategy: BaselineStrategy | None = None
    baseline_window: int | None = Field(default=None, ge=3, le=90)
    threshold_percent: float | None = Field(default=None, gt=0)
    zscore_threshold: float | None = Field(default=None, gt=0)
    check_frequency: CheckFrequency | None = None
    notify_email: bool | None = None
    is_enabled: bool | None = None


class MonitorRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    dataset_id: UUID
    dataset_filename: str
    name: str
    metric_column: str
    aggregation: AggregationOp
    time_column: str
    baseline_strategy: BaselineStrategy
    baseline_window: int
    threshold_percent: float | None
    zscore_threshold: float | None
    check_frequency: CheckFrequency
    notify_email: bool
    is_enabled: bool
    last_checked_at: datetime | None
    last_status: MonitorStatus | None
    last_error: str | None
    created_at: datetime


class AnomalyRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    monitor_id: UUID
    monitor_name: str
    dataset_id: UUID
    metric_column: str
    period_label: str
    observed_value: float
    baseline_value: float | None
    change_percent: float | None
    direction: Direction
    detection_method: DetectionMethod
    severity: Severity
    explanation: str | None
    alert_sent: bool
    alert_sent_at: datetime | None
    alert_error: str | None
    created_at: datetime


class MonitorRunResult(BaseModel):
    status: MonitorStatus
    message: str | None = None
    anomaly: AnomalyRead | None = None

File: frontend/src/types/monitor.ts
import type { AggregationOp } from '@/types/dataset'

export type BaselineStrategy = 'previous_period' | 'moving_average' | 'zscore'
export type CheckFrequency = 'hourly' | 'daily' | 'manual'
export type MonitorStatus = 'NO_ANOMALY' | 'ANOMALY_DETECTED' | 'INSUFFICIENT_DATA' | 'ERROR'
export type Direction = 'increase' | 'decrease'
export type Severity = 'low' | 'medium' | 'high'

export interface Monitor {
  id: string
  dataset_id: string
  dataset_filename: string
  name: string
  metric_column: string
  aggregation: AggregationOp
  time_column: string
  baseline_strategy: BaselineStrategy
  baseline_window: number
  threshold_percent: number | null
  zscore_threshold: number | null
  check_frequency: CheckFrequency
  notify_email: boolean
  is_enabled: boolean
  last_checked_at: string | null
  last_status: MonitorStatus | null
  last_error: string | null
  created_at: string
}

export interface MonitorCreate {
  dataset_id: string
  name: string
  metric_column: string
  aggregation: AggregationOp
  time_column: string
  baseline_strategy: BaselineStrategy
  baseline_window?: number
  threshold_percent?: number | null
  zscore_threshold?: number | null
  check_frequency: CheckFrequency
  notify_email: boolean
}

export interface MonitorUpdate {
  name?: string
  baseline_strategy?: BaselineStrategy
  baseline_window?: number
  threshold_percent?: number | null
  zscore_threshold?: number | null
  check_frequency?: CheckFrequency
  notify_email?: boolean
  is_enabled?: boolean
}

export interface Anomaly {
  id: string
  monitor_id: string
  monitor_name: string
  dataset_id: string
  metric_column: string
  period_label: string
  observed_value: number
  baseline_value: number | null
  change_percent: number | null
  direction: Direction
  detection_method: string
  severity: Severity
  explanation: string | null
  alert_sent: boolean
  alert_sent_at: string | null
  alert_error: string | null
  created_at: string
}

export interface MonitorRunResult {
  status: MonitorStatus
  message: string | null
  anomaly: Anomaly | null
}
